### ISSUE TITLE
[css-color-4] Evaluate static gamut mapping algorithms for oklab/oklch spaces

### ISSUE THREAD

**Issue by ccameron-chromium** [w3c/csswg-drafts#10579]
Separating the action item "ccameron: review gamut mapping algorithms" [from here](https://github.com/w3c/csswg-drafts/issues/9449#issuecomment-2162711081) into a new issue. Tagging in @foolip @LeaVerou @svgeesus for the more limited discussion here (please feel free to include anyone else).

From the linked resolution in the issue, these algorithms are not to be applied to the various RGB spaces. I'd like to scope my comments to just "gamut mapping to apply to oklab/oklch" spaces, since I feel a bit differently about XYZ and Lab spaces (I would want XYZ and Lab to follow RGB, because they are the canonical profile connection spaces used in ICC based color matching, and don't want to mix this discussion with those issues).

The properties that I'd like for the algorithm to apply to just-oklab/oklch are much looser. An iterative algorithm with quadratic convergence is quite fine.

There were lots of algorithms listed in issue 9449, so I'd like to collect a reduced list of candidates here to discuss.

To provide an extremely short list:
* Projecting in Lab to the Rec2020 gamut boundary along constant L
  * This is what's proposed in the spec (modulo termination conditions and iteration method)
  * We can write out the details for an exact solution that quadraticly converges (it's a bit of work, but I can do it if desired)
  * An exact solution for the exact Rec2020 boundary may not be desirable
    * There is some concave nastiness around the blue corner of the gamut that it would be nice to avoid.
    * There is also chrome reduction as discussed in the spec
    * Creating a "nicer-shaped" surface very close to the Rec2020 gamut might be a good solution here
* Projecting in Rec2020 YUV along constant Y
  * This is extremely fast and probably quite good.

**Comment by svgeesus** [comment]
Thanks for starting this.

> * Projecting in Lab to the Rec2020 gamut boundary along constant L

Why Lab and not Oklab, given the hue curvature of Lab (especially in blues, but also in orange to a lesser extent).

I agree that mapping to a smoothed gamut boundary may well be better.

> * Projecting in Rec2020 YUV along constant Y

That is video luma, right? Where Y is computed without bothering to undo gamma-encoding first?

**Comment by ccameron-chromium** [comment]
> Thanks for starting this.
> 
> > * Projecting in Lab to the Rec2020 gamut boundary along constant L
> 
> Why Lab and not Oklab, given the hue curvature of Lab (especially in blues, but also in orange to a lesser extent).

Sorry, that was a typo. I fixed up the original comment.

> I agree that mapping to a smoothed gamut boundary may well be better.
> 
> > * Projecting in Rec2020 YUV along constant Y
> 
> That is video luma, right? Where Y is computed without bothering to undo gamma-encoding first?

In the experiments that I did (which was not very exhaustive), I was doing this in linear space, but that was just because it was easier to plug into the pipeline at that stage.

Another option was to find the nearest point in the L2 sense in YUV space (either linear or gamma encoded). I haven't investigated that thoroughly.

**Comment by svgeesus** [comment]
> Sorry, that was a typo. I fixed up the original comment.

Ah okay, makes sense.

> I was doing this in linear space

That makes the calculated lightness accurate, so good. It doesn't make the UV plane perceptually uniform, of course.

**Comment by svgeesus** [comment]
> There is some concave nastiness around the blue corner of the gamut that it would be nice to avoid.

This is clearly seen in [CSS Color 4, Figure 2](https://drafts.csswg.org/css-color-4/#fig-three-gamuts).

![sRGB-DisplayP3-rec2020-in-Oklab](https://github.com/user-attachments/assets/e12b0c9a-15e3-4c96-ad34-eddcda241450)

**Comment by svgeesus** [comment]
Looking again at [](https://apps.colorjs.io/gamut-mapping/?color=oklch%2890%25+.1+120%29) but this time for desaturated colors which are actually in gamut, I see unacceptable results for the `Scale` and `ScaleLH` options:

![Image](https://github.com/user-attachments/assets/9de014a5-8912-47e4-ac81-744ba04d723e)

![Image](https://github.com/user-attachments/assets/cd3ed5c0-6a88-46ee-abd5-23db2ab86433)

![Image](https://github.com/user-attachments/assets/04730ed2-7a80-4720-840f-fa90c82b984a)

**Comment by svgeesus** [comment]
Taking the results for "CSS Rec2020" as the consensus target (CSS gamut map to the BT.2020 gamut, then clip from there to the display) the four contenders are thus:

- Chromium
- Björn Ottosson 
- Raytrace
- Edge Seeker

**Comment by svgeesus** [comment]
Looking now at out of gamut colors (including colors even outside ProPhoto) all four give similar results to "CSS Rec2020" and often slightly better results.

For example, for [oklch(80% 109% 311)](https://apps.colorjs.io/gamut-mapping/?color=oklch%2880%25+109%25+311%29) CSS Rec2020 has a ΔH of 4.3, Chromium ΔH 1.7, Edge Seeker ΔH 0.03 while Björn Ottosson and Raytrace both have ΔH 0, which is ideal.

![Image](https://github.com/user-attachments/assets/0fc3bba0-109a-41aa-b9a8-559cf5dcdfb0)

![Image](https://github.com/user-attachments/assets/f60df86c-62ec-42a4-9646-46fe8236ce92)

I think the next task is to capture timing information on those four.

**Comment by svgeesus** [comment]
Looking at gamut mapping each position in a [361-stop gradient from `oklch(90% .6 250)` to `oklch(40% .3 20)`](https://apps.colorjs.io/gamut-mapping/gradients?from=oklch%2890%25+.6+250%29&to=oklch%2840%25+.3+20%29) (set max DeltaE to 3, to get lots of stops) the timings in that case, sorted from slowest to fastest:

- CSS Rec2020 **21.3ms**
- CSS **17ms**
- Raytrace **12.9ms**
- Scale LH **9.1ms**
- Björn Ottosson **6.4ms**
- Chromium **5.5m**s
- Edge Seeker **3.7ms**
- Clip **2.9ms**

**Comment by svgeesus** [comment]
Calculating mean runtime over 100 iterations gives similar speed ranking:

- CSS **39.285**ms
- CSS Rec2020 **37.438**ms
- Raytrace **25.618**ms
- Scale LH **21.532**ms
- Björn Ottosson **12.361**ms
- Chromium **10.203**ms
- Edge Seeker **7.488**ms
- Clip **5.062**ms

The [code for edge seeker](https://github.com/color-js/apps/blob/main/gamut-mapping/edge-seeker/makeEdgeSeeker.js)  calculates the maximum chroma for a given lightness and hue. The gamut slice is approximated with a linear segment for the lower (dark) part and an arc for the upper (light) part. This implementation populates a 400-element LUT.

**Comment by jamesnw** [comment]
I generated colors in the XYZ space, and for each color outside of Rec2020, I applied the GMA, and then clipped to P3 and sRGB to approximate what color might be displayed.  I then calculated average deltas for each channel in OkLCh, taking the absolute value since I was interested in magnitude and not directionality. A drawback is that this obscures how the GMAs behave in different ranges, as well as how much variation there is.

| Method | Destination Gamut | ΔL | ΔC | Δh  | ΔDelta2000 |
|--------|--------|--------|--------|--------|--------|
| Edge Seeker p3 | p3 | 0.0004 | 0.3101 | 0.0007 | 17.4262 |
| Edge Seeker  p3| srgb | 1.4090 | 0.3329 | 2.2804 | 19.2775 |
| Edge Seeker Rec2020 | p3 | 2.1310 | 0.2910 | 2.9192 | 16.2893 |
| Edge Seeker Rec2020 | srgb |  3.3636 | 0.3172 |  5.2885 | 18.7126 |
| Björn Ottosson | p3 | 0.000008 | 0.3101 | 0.00002 | 17.4203 |
| Björn Ottosson | srgb | 1.4105 | 0.3328 | 2.2829 | 19.2729 |
| Chromium | p3 | 2.3165 | 0.2858 | 3.2774 | 15.8680 |
| Chromium | srgb | 3.4980 | 0.3139 | 5.6398 | 18.4112 |
| Raytrace | p3 |  0.00002 | 0.3101 | 0.0025 | 17.4188 |
| Raytrace | srgb | 1.4106 | 0.3328 | 2.2853 | 19.2720 |
| Clipping | p3 | 7.2 | 0.244 | 8.72074 | 16.607 |


Scales: 
  * L = 0-100  (So a 1 would be equivalent to 0.01 or 1%)
  * C = 0-0.4 (Most of the mapped colors had very high chroma to start with, so we're seeing large shifts here, as expected)
  * h = 0-360
 
Some notes-
* All are improvements over Clipping
* Edge Seeker performs well. Given its speed and performance, I would support this being the method for getting to rec2020.
* Björn Ottosson performs much better than Edge Seeker when clipping to p3, but the advantage isn't there when clipping to srgb (Although the implementation is using p3 coefficients, and we could use separate sRGB coefficients, I think?). Is this performance boost worth the almost 2x longer compute time?
* ~~I omitted Raytrace since it appears to be slower~~ (added)
* The implementations of both Edge Seeker and Björn Ottosson target a displayP3 output in how they are calculated- would we want to target a rec2020 output? I assume that would negatively impact their performance.
  *  UPDATE June 10 - Added Edge Seeker with rec2020-based LUT.

**Comment by facelessuser** [comment]
> I omitted Raytrace since it appears to be slower.

It should be noted that you can decrease the raytrace iteration from 4 to 3 and reduce its speed and still likely have pretty good results, though I wouldn't go lower than 3. I'd suspect with high loads that Björn's method would still prove overall faster.

Raytrace is decent as a generic approach for various RGB gamuts as you don't need to hard code color space specific coefficients, but since it sounds like the plan is to just target a single gamut, Rec. 2020, and clip to everything else, that doesn't matter so much and something like like Björn Ottosson's approach is definitely faster, and Edge Seeker even more so if storing a LUT is acceptable.

**Comment by facelessuser** [comment]
EDIT: This test is done with the full 4 iterations of raytrace, no reduction.

Okay, one thing that was bothering me about the Color.js testing of GMAs is that it isn't 100% fair as some algorithms are getting hit heavier by certain API calls of Color.js.

Björn and Edgeseeker bypass the Color.js conversion pipeline which has a lot of overhead. For instance, Björn's converts to linear RGB outside of of the Color.js API.

So optimizing Raytrace to do the same, we actually get faster results than Björn's.

Forcing 1000 gradient steps, ray-trace is shown to be noticeably faster.

![Image](https://github.com/user-attachments/assets/a880c9dc-fecf-4bd7-89cf-62ca04057088)

Here's the optimized preview: https://deploy-preview-21--color-apps.netlify.app/gamut-mapping/gradients?from=oklch%2890%25+.4+250%29&to=oklch%2840%25+.1+20%29

The optimizations can be seen here: https://github.com/color-js/apps/pull/21.

I imagine even the CSS approach could be sped up doing something similar.

**Comment by jamesnw** [comment]
> It should be noted that you can decrease the raytrace iteration from 4 to 3 and reduce its speed and still likely have pretty good results, though I wouldn't go lower than 3. I'd suspect with high loads that Björn's method would still prove overall faster.

I added in Raytrace, and it looks like it outperforms Edge Seeker on Lightness, but not hue. It also outperforms Björn's method on Delta2000.

**Comment by facelessuser** [comment]
> I added in Raytrace, and it looks like it outperforms Edge Seeker on Lightness, but not hue. It also outperforms Björn's method on Delta2000.

Yep, that's about what I would expect. It'll get hue reasonably close, but it can't beat a LUT 🙂.

**Comment by facelessuser** [comment]
You can actually get better results with Raytrace by increasing the iterations at the cost of the time it takes to perform another iteration. So it can be tuned a bit to get tighter results. I picked 4 iterations as a decent middle ground. I think at Rec. 2020 width you get a little more hue shift, but it was deemed within acceptable limits. You could probably even increase iterations to 5 and potentially still be competitive. Just something to note.

**Comment by facelessuser** [comment]
Looking over the data, if holding lightness constant is a strong desire, then excluding the Chromium approach, I think the difference in deltas are negligible and the human eye isn't going to be able to pick out the difference, so generally, they all seem pretty good. So, that just leaves speed, and if that is the deciding metric, and storing a LUT in memory is acceptable, then I agree that the LUT based approach clearly wins.

**Comment by jamesnw** [comment]
I realized I could quickly alter Edge Seeker to generate a rec2020-based LUT, and as I suspected, the performance dropped significantly. I've added it to the [table above](https://github.com/w3c/csswg-drafts/issues/10579#issuecomment-2954505854) as `Edge Seeker rec2020`. It roughly matches the Chromium approach. 

I didn't see a quick way to switch RayTrace or Björn to get to the rec2020 boundary, but would suspect similar performance differences. The `clip to display gamut` step is likely to be the step with the largest non-Chroma change, so I still agree that the LUT/Edge Seeker algorithm is a good path forward.

**Comment by facelessuser** [comment]
@jamesnw I would expect results in the same ballpark for the others as well. If you want to directly test, Björn's approach and raytrace can easily be adjusted to do rec2020 as well if your interested.

For Björn's you need new Rec. 2020 matrices and coefficients (you can update the variable names). Also, update any conversions using the Color.js API to reference Rec.2020 opposed to P3.

```js
               lmsToP3Linear: [
			[2.1399067304346517, -1.2463894937606181, 0.10648276332596672],
			[-0.8847358357577675, 2.1632309383612007, -0.27849510260343346],
			[-0.04857374640044416, -0.454503149714096, 1.5030768961145402]
		],
		P3Coeff: [
			// Red
			[
				// Limit
				[-1.36834899, -0.46664773],
				// `Kn` coefficients
				[1.2572445, 1.71580176, 0.5648733,  0.79507316, 0.58716363],
			],
			// Green
			[
				// Limit
				[2.01150796, -2.0379096],
				// `Kn` coefficients
				[0.74087755, -0.4586733, 0.08182977, 0.12598705, -0.14570327],
			],
			// Blue
			[
				// Limit
				[0.06454093, 2.29709336],
				// `Kn` coefficients
				[1.36920484e+00, -1.64666735e-02, -1.14197870e+00, -5.01064768e-01, 1.19905985e-03],
			],
		]
```

For ray trace, you just need to update the RGB to LMS and LMS to RGB matrices. Also, update any conversions using the Color.js API to reference Rec.2020 and/or Linear Rec. 2020 opposed to P3.

```js
		oklchToLinearRGB (lch) {
			// Convert from Oklab to linear RGB.
			//
			// Can be any gamut as long as `lmsToRgb` is a matrix
			// that transform the LMS values to the linear RGB space.

			let c = lch[1];
			let h = lch[2];
			// to lab
			let result = [
				lch[0],
				c * Math.cos((h * Math.PI) / 180),
				c * Math.sin((h * Math.PI) / 180)
			];

			// To LMS
			util.multiply_v3_m3x3(
				result,
				[
					[ 1.0000000000000000,  0.3963377773761749,  0.2158037573099136 ],
					[ 1.0000000000000000, -0.1055613458156586, -0.0638541728258133 ],
					[ 1.0000000000000000, -0.0894841775298119, -1.2914855480194092 ]
				],
				result
			);
			result[0] = result[0] ** 3;
			result[1] = result[1] ** 3;
			result[2] = result[2] ** 3;

			// To RGB
			util.multiply_v3_m3x3(
				result,
				[
					[2.1399067304346517, -1.2463894937606181, 0.10648276332596672],
					[-0.8847358357577675, 2.1632309383612007, -0.27849510260343346],
					[-0.04857374640044416, -0.454503149714096, 1.5030768961145402]
				],
				result
			);
			return result;
		},

		LinearRGBtoOklch (rgb) {
			// Convert from Oklch to linear RGB.
			//
			// Can be any gamut as long as `lmsToRgb` is a matrix
			// that transform the LMS values to the linear RGB space.

			// To LMS
			let result = util.multiply_v3_m3x3(
				rgb,
				[
					[0.6167557848654442, 0.3601984012264634, 0.023045813908092294],
					[0.2651330593926367, 0.6358393720678492, 0.0990275685395141],
					[0.10010262952034829, 0.20390652261661446, 0.6959908478630372]
				]
			);

			result[0] = Math.cbrt(result[0]);
			result[1] = Math.cbrt(result[1]);
			result[2] = Math.cbrt(result[2]);

			util.multiply_v3_m3x3(
				result,
				[
					[ 0.2104542683093140,  0.7936177747023054, -0.0040720430116193 ],
					[ 1.9779985324311684, -2.4285922420485799,  0.4505937096174110 ],
					[ 0.0259040424655478,  0.7827717124575296, -0.8086757549230774 ]
				],
				result
			);

			let a = result[1];
			let b = result[2];
			return [
				result[0],
				Math.sqrt(a ** 2 + b ** 2),
				constrainAngle((Math.atan2(b, a) * 180) / Math.PI)
			];
		},

```

**Comment by jamesnw** [comment]
I added those, as well as my comparison code in this PR- https://github.com/color-js/apps/pull/22. 

You can compare here- https://deploy-preview-22--color-apps.netlify.app/gamut-mapping/compare/

**Comment by facelessuser** [comment]
Yep, results are about what I expected, roughly in the same ballpark. Chromium is slightly different due to its rougher approximation, but generally in the same range.

**Comment by svgeesus** [comment]
> Looking over the data, if holding lightness constant is a strong desire, then excluding the Chromium approach, I think the difference in deltas are negligible and the human eye isn't going to be able to pick out the difference, so generally, they all seem pretty good. 

More or less. The ones that map to BT.2020 and then clip from there have a visually noticeable shift towards the saturated secondary colors, but the ones that go directly to Display P3 all look visually the same.

> So, that just leaves speed, and if that is the deciding metric, and storing a LUT in memory is acceptable, then I agree that the LUT based approach clearly wins.

I agree (and should have said so explicitly earlier). Notably, the speed difference between the LUT approach and simple clip is very small.

**Comment by mirisuzanne** [comment]
What's the next step needed here? Do we need input from vendors on the LUT approach? Do we still need to select one of the methods, and add it to spec? Something else?

**Comment by svgeesus** [comment]
> Separating the action item "ccameron: review gamut mapping algorithms" [from here](https://github.com/w3c/csswg-drafts/issues/9449#issuecomment-2162711081) into a new issue.

The next step, now that others have

- proposed algorithms
- implemented them
- evaluated the quality of the result
- evaluated the speed of each
- chosen a reduced list of candidates, as requested

is that the original action item  for @ccameron-chromium gets done.

But I would certainly value input from other browser vendors, too.

And once one (or more?) are selected, I will add them to the spec.

**Comment by ccameron-chromium** [comment]
Given the scope (that these are applied to `oklab` and `oklch` spaces, and are targeting a static gamut), these are all good options to Chromium.

The simpler the better, and the more likely to get picked up in implementation implemented. The "polyhedral thing" that I implemented in Chromium behind a flag is, in retrospect, too complicated.

A simple operation operation (eg, projecting to the rec2020 gamut in linear space) would probably be my personal favorite.

**Comment by svgeesus** [comment]
> A simple operation operation (eg, projecting to the rec2020 gamut in linear space) would probably be my personal favorite.

Since linear-light spaces are not perceptually uniform, it would need to be shown that the results from that operation are good enough (and that it is significantly faster than the best of the options presented here).

> these are all good options to Chromium.

It would be great to hear from the Gecko and WebKit point of view as well.

**Comment by svgeesus** [comment]
Proposed Resolution: Add EdgeSeeker and Raytrace GMA (to rec2020) to CSS Color 4

@astearns I think this one could be async

**Comment by LeaVerou** [comment]
Adding some data I recently observed and I haven't seen brought up before. It appears that clipping produces a *much* worse color for higher lightnesses whereas it is much more acceptable for medium to low lightness even for fairly substantial clips.

Demo: https://codepen.io/leaverou/pen/KwdMVVG
Tweak the color on the left and see how different L values render, both raw in the browser (currently clipped) and gamut mapped to P3 and sRGB.
The very light colors suffer the most, even when the original color (before the L adjustment) was only slightly out of gamut:

<img width="1799" height="530" alt="Image" src="https://github.com/user-attachments/assets/677af1f8-6434-4fa9-921e-3fea37841aaa" />

This comes up a lot with dynamic tint generation (what originally prompted this).

Whereas for the rest, you need to be much more significantly OOG to see a significant difference, so this approach would work fine:

<img width="1799" height="529" alt="Image" src="https://github.com/user-attachments/assets/8402c5ad-3c24-48f3-9593-c11f169fd260" />

I wonder if rather than having a fixed "gamut map to REC 2020 and then clip" algorithm we could have a fast lightness test and take different code paths, using clipping more for darker colors and more gamut mapping for very light colors.

Edit: NVM perhaps. The issue is probably that after the L adjustment the lighter colors are more wildly OOG, so gamut mapping them to Rec.2020 first is fine. I added a scale that does just that:

<img width="1797" height="695" alt="Image" src="https://github.com/user-attachments/assets/d1310a47-22de-4393-9152-f755aab331f6" />

**Comment by facelessuser** [comment]
> Adding some data I recently observed and I haven't seen brought up before. It appears that clipping produces a much worse color for higher lightnesses whereas it is much more acceptable for medium to low lightness even for fairly substantial clips.

This is something I've brought up before. Clipping and gamut mapping that preserves lightness and hue are good for different things. The "gamut map to Rec. 2020 and then clip" was proposed to favor images, but not for applications like you are demonstrating. Preserving lightness and hue throughout the entire gamut mapping process obviously serves your use case better.

This is one of the reasons I've thought that gamut mapping images and gamut mapping colors in CSS don't necessarily need to use the same approach, or at the very least should give you a choice. If coming from the approach that only one algorithm is desired, I think the "gamut map to Rec. 2020 and then clip" was supposed to be an approach to kind of meet in the middle; gamut map extreme colors, but then clip to make sure images look good.

**Comment by mirisuzanne** [comment]
It's not clear to me if this resolution only applies to oklab/oklch colors?

**Comment by jamesnw** [comment]
> It's not clear to me if this resolution only applies to oklab/oklch colors?

It's also unclear to me, but to gather the relevant info- 

The [initial resolution](https://github.com/w3c/csswg-drafts/issues/9449#issuecomment-2162711081) was "Gamut map all device independent color spaces" and "RGB color spaces (srgb, display-p3, rec2020, a98rgb) do not gamut map."

Device independent would include Lab, LCH, Oklab, and OKLch. XYZ spaces aren't specifically mentioned in the resolution. @ccameron-chromium suggests [limiting this change](https://github.com/w3c/csswg-drafts/issues/10579#issue-2411507542) to just Oklab and OKLch.  

In addition to specifying the impacted spaces, we should specify when gamut mapping is triggered in a case like this-

```css
a {
  background: linear-gradient(in rec2020, oklch(70% 0.5 340), oklch(90% 0.5 200));
}
b {
  background: linear-gradient(in oklch, yellow, blue);
}
```

I would expect `a` to be clipped and `b` to be gamut mapped where it goes out of rec2020- in other words, the interpolation space, not the space of the colors being interpolated, would be the deciding factor.

**Comment by svgeesus** [comment]
> It's not clear to me if this resolution only applies to oklab/oklch colors?

I believe that was explicitly the intent (which will require spec changes, because currently gamut mapping is invoked whenever a color value is outside the gamut of the display device).

> Device independent would include Lab, LCH, Oklab, and OKLch. XYZ spaces aren't specifically mentioned in the resolution. [@ccameron-chromium](https://github.com/ccameron-chromium) suggests [limiting this change](https://github.com/w3c/csswg-drafts/issues/10579#issue-2411507542) to just Oklab and OKLch.

Right. I believe that the motivations were:

- ICC profile handling relies on CIE Lab (D50) and CIE XYZ (D50) so those should not be altered
- OkLab (and OkLCH) are optimized for hue linearity over moderately wide-gamut colors, but were not designed for (and give very odd results) for ultra-wide gamut and especially, for imaginary colors

Thus, the agreement that such colors should be gamut mapped into the BT.2020 gamut, which is wide gamut but does not have any imaginary colors.

It isn't clear to me when this happens and whether the result of such mapping is reflected in the computed value.

**Comment by svgeesus** [comment]
> It isn't clear to me when this happens and whether the result of such mapping is reflected in the computed value.

It still isn't clear, any takers?

**Comment by svgeesus** [comment]
> I would expect `a` to be clipped and `b` to be gamut mapped where it goes out of rec2020- in other words, the interpolation space, not the space of the colors being interpolated, would be the deciding factor.

That seems reasonable, so next question: is it

- the colors being interpolated
- the interpolated result

which is gamut mapped? The first option involves less computation but it is trivial in polar spaces to get an oog result from two in-gamut colors.

**Comment by LeaVerou** [comment]
@ccameron-chromium what is the status on this? Where is it blocked? We're going to hit web compat if we don't ship gamut mapping soon.

**Comment by svgeesus** [comment]
@jamesnw @facelessuser I have committed early drafts or EdgeSeeker and Ray Trace sections into the spec (including a figure of yours, @facelessuser which I request permission to use - please let me know and I will of course remove it if you would rather not).

This commit gives implementations a free choice of the three algorithms but they must use one of them whenever CSS says a GM is required. Which is currently when mapping to a display.

We also have a resolution to use GMA to rec2020 to tame `Oklab` and `Oklch`, which is not yet in the spec and **will affect computed values.**

The main blocker from my POV is generating pseudocode versions of these algorithms; they are faster than the binary search with local MINDE but the code is somewhat more complex and might not lend itself to pseudocode. But these algorithms need to be implementable from the spec.

I didn't add timing results but did add general mentions of quality/speed/memory tradeoffs.

**Comment by facelessuser** [comment]
@svgeesus You are free to use the image.

**Comment by jamesnw** [comment]
@svgeesus While I wish I could claim credit, Edgeseeker was contributed by @ardov in https://github.com/color-js/color.js/pull/448.

**Comment by facelessuser** [comment]
This can help as a starting point for pseudo-code for the raytrace algorithm.

To ensure better performance, it is assumed that gamut mapping between the supported CSS gamuts (sRGB, Display P3, and
Rec. 2020) is performed with direct conversions between the linear RGB space to Oklab using matrices that combine the
translation of linear RGB to XYZ and XYZ to LMS. These can be done in separate steps, but combining these
calculations in one matrix can reduce the number of multiplications.

RGB gamuts are assumed to be simple rectangular prisms, positioned at [0, 0, 0] extending in the positive direction.
While there is no requirement that the gamut be limited to [1, 1, 1], we will assume SDR lightness, and assume max
values of [1, 1, 1] equalling OkLCh [1, 0, 0]. Colors that exceed the minimum OkLCh lightness of 0 and 1 would exceed
this range and be clamped to black or white, respectively. The important thing to note is that the lightness check
corresponds with the RGB constraints.

The ray trace algorithm is based on the slab method (https://en.wikipedia.org/wiki/Slab_method).

1. Assume `start` to be an anchor within the RGB gamut and `end` to be the color we wish to locate on the RGB gamut surface.
2. Assume `bmin` and `bmax` to be arrays with the gamut's lower and upper bounds, respectively. These could be collapsed to a single value variable, as practically all component min and max values will be the same for all color channels.
3. Set `tfar` to infinity or some very large number.
4. Set `tnear` to -infinity or some very large, negative number.
5. Let `direction` be a 3-element array.
6. For `(i = 0; i < 3; i++)`:
    1. `a = start[i]`
    2. `b = end[i]`
    3. `d = b - a`
    4. store `d` in `direction[i]`
    5. if `d != 0`:
        6. `inv_d = 1 / d`
        7. `t1 = (bmin[i] - a) * inv_d`
        8. `t2 = (bmax[i] - a) * inv_d`
        9. `tnear = max(min(t1, t2), tnear)`
        10. `tfar = min(max(t1, t2), tfar)`
    6. else if `a < bmin[i]` or `a > bmax[i]` return an indictation that no intersection was found
7. If `tnear > tfar` or `tfar < 0`, return an indication that no intersection was found
8. If `tnear < 0`, set tnear to tfar, favoring the first intersection in the direction `start` -> `end`
9. If `tnear` is infinite (or matches the initial very large value), return an indication that no intersection was found.
10. Return the intersection by interpolation such that each index `i` in the intersection corresponds to `start[i] + direction[i] * tnear`.

To perform the ray trace gamut mapping approach:

1. If `orign` is within the target CSS gamut, return `orign`.
2. Convert `origin` to OkLCh as `oklch_orign`.
3. If `oklch_origin` lightness exceeds 1, white is returned.
4. If `oklch_origin` lightness is below 0, black is returned.
5. Lightness and hue are saved for later use as `l_origin` and `h_origin`.
6. Using `l_origin`, 0, and `h_origin` create an achromatic OkLCh color and convert it to the linear RGB gamut as `anchor`.
7. `oklch_origin` is then converted to the linear RGB gamut as `rgb_origin`
8. Set a lower threshold of 1e-6 as `low`. It is assumed the minimum value is 0 and that all channels have the same minimum. The value should be small relative to the unit type. 64 bit could easily be as small as 1e-14, but 1e-6 is fine.
9. Set a higher threshold of `1 - low` as high. 1 represents the max channel value, and it is assumed all channels have the same max.
10. Store the current `rgb_origin` as `last`
11. For `(i = 0; i < 4; i++)`:
    1. If `i > 0`:
        1. Convert `rgb_origin` to OkLCh and set lightness and hue to `l_origin` and `h_origin` respectively. This places the current color back on the chroma reduction curve if it has deviated.
        2. Convert back to linear RGB and store in `rgb_origin`
    2. Cast a ray from the `anchor` to `rgb_origin` and save the gamut intersection as `intersection`.
    3. If `intersection` was not found, update `rgb_origin` with `last` and exit the loop.
    4. If `i > 0` and each component in `rgb_origin` is between the `low` and `high` threshold, update `anchor` with `rgb_origin`. This means `rgb_origin` is below the gamut surface, use it as an anchor closer to the gamut surface.
    5. Update `rgb_origin` and `last` with `intersection`.
12. Clip `rgb_origin` to gamut, trimming off any noise.
13. Convert `rgb_origin` back to the color space of `origin` and update `origin`.

**Comment by facelessuser** [comment]
Updated a few typos, added more clarification, and fixed a missed check in the algorithm. Hopefully, everything is accurate now, but feel free to ask questions if anything is unclear.

**Comment by svgeesus** [comment]
Will do but thanks, this is great!

**Comment by svgeesus** [comment]
> 2\. b = start[i]

That should be b = end[i], right, same as [this code](https://github.com/color-js/apps/blob/2c7346dd00855f7b82eb4c7527355a09b84beeb3/gamut-mapping/methods.js#L457)

**Comment by facelessuser** [comment]
Yep, typo on my part.

**Comment by svgeesus** [comment]
> * Clip `rgb_origin` to gamut, trimming off any noise.
> 
> * Convert `rgb_origin` back to the color space of `origin` and update `origin`.

Should those be in the opposite order, ie convert then clip, because the conversion might also introduce some floating point noise?

**Comment by facelessuser** [comment]
What if origin is in something like Oklab and fitting within linear sRGB? You want to clip while still in linear sRGB. I guess if the final color representation is in the same gamut as the fitting is done in, you could clip after 🤷🏻. I guess you can make the call based on how it's specifically used in CSS.

**Comment by svgeesus** [comment]
If you get a chance, a quick review of what is in the spec now would be super helpful. It isn't a straight cut and paste: I tried to make it in the WHATWG spec pseudocode style, in general, and to say things the same way as the binary search GMA pseudocode, in particular.

https://drafts.csswg.org/css-color-4/#pseudo-raytrace

**Comment by facelessuser** [comment]
@svgeesus 

## GMA algorithm

2\) seems to have a comment at the end that needs to be removed `!-- constrain to SDR lightness range, which gamut maps to black or white -->`

5\) has `and h_origin.` at the end which should be removed

## Ray trace algorithm

6\) `if (tnear > tfar or tfar < 0` Seems it is missing a `)`

9\) `for (i =0; i lt; 3; i++):` I think you mean t `i < 3`, maybe an HTML enitity typo?

9\) and 10\) seem to reuse `end` to store the intersection. This should probably not be done, assuming `end` is the current RGB point and not a copy. Once we've got the intersection, we want to use that RGB point (not the intersection) to update the anchor, but only after we've used the old anchor. If we overwrite the RGB point with the intersection, we'll end up not having the anchor update properly. Depending on the language, this could be handled a number of ways, `end` is a copy of the data, you could pass in a reference to an array that will be filled with the intersection, you could allocate and return a new array as well, etc. I think this should be written in a way that does not lead people to believe they could overwrite the original data.

Outside of that, I think everything seems correct. We'll see if/when people start to try and implement.

**Comment by facelessuser** [comment]
@svgeesus Some notes on the description:

> The color to be mapped is first converted to Oklch, and then the achromatic version of that color is generated, which will be the neutrax axis anchor. These two colors are then converted to the linear-light version of the target RGB space.

I think you meant "neutral" not "neutrax"

> This process a maximum of three times, each time finding a better, closer color on the path. Finally, simple clipping is used to account for floating point math errors.

I think you meant "This process _is performed_ a maximum of three _additional_ times"

**Comment by svgeesus** [comment]
> 9) and 10) seem to reuse `end` to store the intersection. This should probably not be done, assuming `end` is the current RGB point and not a copy. Once we've got the intersection, we want to use that RGB point (not the intersection) to update the anchor, but only after we've used the old anchor. If we overwrite the RGB point with the intersection, we'll end up not having the anchor update properly. Depending on the language, this could be handled a number of ways, `end` is a copy of the data, you could pass in a reference to an array that will be filled with the intersection, you could allocate and return a new array as well, etc. I think this should be written in a way that does not lead people to believe they could overwrite the original data.

Ah yes, it should not clobber `end` because it is called in a loop. Like [here in color.js](https://github.com/color-js/apps/blob/2c7346dd00855f7b82eb4c7527355a09b84beeb3/gamut-mapping/methods.js#L493-L498). Good catch, fixed.

**Comment by facelessuser** [comment]
Yep, it looks good now.

**Comment by facelessuser** [comment]
@svgeesus I'd like to make one small change to the algorithm. Seeing the ray trace algorithm made it into the spec, and even though I've been using the algorithm for a year and half, I started putting it through some extensive testing, just to make sure. This exposed a rare issue that can occur, which is easy for us to fix.

Change this:

```
8. if tnear is infinite (or matches the initial very large value)
   - return INTERSECTION NOT FOUND
```

to 

```
8. if tnear is greater than 10
   - return INTERSECTION NOT FOUND
```

Basically, when this value is getting large, we are really close to the gamut surface. Infinity is basically a perfect scenario, but any large value can indicate we are close enough to the surface to consider we are done. If we accept too large of values, it can throw off our intersection calculation.

Just to illustrate the issue. Consider this value `color(srgb 1.0000000000000004 0.19694944045886786 0.026717808177518357)`, which is very close to the surface. On the fourth iteration, we get a very large `tnear` that is not infinity. This causes our intersection to be calculated incorrectly.

<img width="800" height="800" alt="Image" src="https://github.com/user-attachments/assets/7bd06be1-cdef-4eda-bd9a-ec573d93a6ae" />

With the fix in place, it clears up the issue as we now just stop when the value is too big.

<img width="800" height="800" alt="Image" src="https://github.com/user-attachments/assets/9cf9c8c7-cd83-4ed5-ae85-060a82c11b57" />

**Comment by svgeesus** [comment]
@facelessuser so presumably Coloraide and color.js should be patched for this change, too?

**Comment by facelessuser** [comment]
@svgeesus I've already patched it over in the color.js reference. I've patched it in ColorAide, but haven't released it yet.

**Comment by svgeesus** [comment]
> I've already patched it over in the color.js reference.

yes, I see
 - https://github.com/color-js/apps/commit/3c6ef577a8468c6fa90341ede97bd9362af56ed2

**Comment by jamesnw** [comment]
I opened https://github.com/color-js/color.js/issues/713 to port EdgeSeeker and Ray Trace into the color.js package, as they are currently just in the Color.js/apps playground.

**Comment by facelessuser** [comment]
@svgeesus As hard as it is to say, my earlier fix was dumb. While it "works" I jumped too quickly with a Band-Aid. I've created a PR #13416. That shows a proper fix.

Long story short, this is a numerical instability issue due to floating-point numbers. Just artificially clamping `tnear` is not the correct approach. The real solution is ensuring the direction is not too small (greater than some epsilon suitable for the unit type).

Anyway, it looks like I don't meet some criteria in PR due to affiliation, but that at least shows what it is that should be done.

**Comment by facelessuser** [comment]
Color.js is [updated](https://github.com/color-js/apps/commit/dc9a2cc036884e8b0be257a617b1ff1130bb8f13), and ColorAide has been released with the updates and now does gamut mapping with the ray trace approach using OkLCh.

The reason the recent issue went undetected for so long was simply that the threshold of `1e-6` prevented the anchor from getting too close to the target color most of the time. When I was stress testing over the last couple of days, I made that threshold even smaller, which exposed the numerical instability as the direction approached zero.

`tnear` is allowed to be very big, and should be at times, but when the direction is smaller than the floating point precision, `tnear` can be large and nonsense. Only because we are often working with small values that do not trigger this case, `tnear` is usually relatively small. That was my mistake, and if I had spent just a little longer thinking about it, I would have come to that conclusion the first time.

**Comment by svgeesus** [comment]
Thanks for the explanation, and your attention to detail is greatly appreciated.

**Comment by jparnzen** [comment]
@svgeesus, @facelessuser : I think there may be a logic error near the end of the raytrace pseudocode in the 2026-02-27 draft.

At `11. set clipped to clip(current)`, `current` doesn't exist.

I think in the 2026-02-02 draft, that step was:
```
let `clipped` be `origin_rgb` clipped to gamut (components in the range 0 to 1), thus trimming off any noise due to floating point inaccuracy
```

There is a `current_OkLCh` that looks to be defined locally to the loop in the logic, but `origin_rgb` gets based on it. So should `current` still be `origin_rgb`?

**Comment by jparnzen** [comment]
One more possible error:

The fifth bullet comparison under `cast a ray`, under line item 5 (`if abs(d) < 1E-12`) should be greater-than and not less-than.

I've confirmed this is greater-than in both the colorjs.io gamut mapping playground raytrace code and in the ColorAide raytrace code. With less-than, the enclosed logic doesn't seem to execute, or rarely will.

Aside: what's the basic concept for the value of 1e-12? Is it an epsilon to account for floating point fudging?

**Comment by facelessuser** [comment]
@jparnzen Yep, I think the sign is backwards.

As to why it is needed? Basically, when the value gets too small, you are just dealing with floating-point noise. When you divide by these small values later, you get very, very big values that cause things to break. Basically, it's needed for numerical stability. This is a normal issue that can occur in algorithms sometimes that need to avoid zero, as we do in this algorithm. Depending on what you do after, sometimes the algorithm can tolerate such values, but in this case, it cannot.

> There is a current_OkLCh that looks to be defined locally to the loop in the logic, but origin_rgb gets based on it. So should current still be origin_rgb?

I'll have to take a closer look tomorrow, but it's likely that if it doesn't match the reference implementation in Color.js, it probably needs updating. Color.js _should_ follow the algorithm pretty close, except for one step to sanely handle some HDR cases, as it handles more than what is defined in the CSS level 4 spec.

**Comment by facelessuser** [comment]
Actually, I took a look tonight.

> So should current still be origin_rgb?

Yes, `current` is ambiguous, and `origin_rgb` is probably appropriate as it gets converted to the destination and then gets clipped.

**Comment by jparnzen** [comment]
I hope I'm not barging into the discussion by letting you all know this, but Firefox Nightly v153 has 2 of the 3 GMAs (local MINDE and raytracing) implemented for testing and review. Users can set the GMA under `about:config` by setting `layout.css.gamut-mapping.method` to 0 (default, simple clipping), 1 (local MINDE), and 2 (raytracing).

If all goes well, this should graduate to full GA with v153. If anyone notices any issues or has improvements, please file a bug in Firefox's bugzilla.

Many thanks to you all for your hard work on these standards and algorithms! Once edgeseeker is added to the draft, I'll submit a new patch to the Firefox team to get it added (I've been picking at it here, translating the javascript to rust, so if I can help with anything here, please let me know, I'm glad to help!).

**Comment by jamesnw** [comment]
@jparnzen Thanks for implementing this! I did some light kicking of tires, and filed https://bugzilla.mozilla.org/show_bug.cgi?id=2046279 with some oddities I observed.

**Comment by facelessuser** [comment]
It's nice seeing these starting to get implemented. It does seem there is something a little off about the Raytrace one, though. It'll be nice to figure out exactly what that is and see if it was just some confusion in the spec or something else that needs clarification.

**Comment by facelessuser** [comment]
I was looking at the implementation here: https://github.com/mozilla-firefox/firefox/blob/main/servo/components/style/color/gamut/raytrace.rs.

I think I may know what the issue is. The Color.js implementation is using float 64, and that is also what ColorAide (what I developed it under) is using, so threshold values were selected, assuming float 64. The raytrace threshold of 1e-12 is likely too large for float 32 and needs to be adjusted to be smaller. You have about 15 - 16 decimals of precision in float 64, but only 6 - 7 in float 32.

I imagine the 1e-6 is probably still okay for anchor distance from the gamut surface. I suspect if Firefox were to lower the raytrace threshold to work within float 32 (maybe 1e-6 or 1e-5), the Raytrace results will improve, assuming I'm not missing something.

If this ends up being the case, we should add a note in the spec about good 32-bit suggestions.

**Comment by jparnzen** [comment]
@jamesnw , @facelessuser : Thank you both for taking a look and finding/filing some issues! I'll go over the Firefox code in the morning to do more testing, and give the loosened f32 epsilon a try to see how it affects results. I'll let you know what I find, and will check with emilio and tlouw at Mozilla about getting issues fixed.

**Comment by jparnzen** [comment]
@facelessuser, @jamesnw : it looks like the `cast_ray` comparison epsilon/threshold value for f32s is the culprit. I changed it from `1e-12` to `1e-6`, recompiled Nightly, and @jamesnw's bugzilla tests look to be passing now--no weird neon green line in the one gradient, and no spurious purple bars in the other gradient. Same with the gradient GM playground link. So the precision seems to be the source, and I'll put together a new patch for Firefox to update it. Let me know if you'd like me to create an issue ticket to track it for the spec.

One thing I noticed in the gradient GM playground here that I'm wondering if either of you see on your sides when using Nightly: when enabling either local MINDE or the fixed raytracing GMA, the Chromium GMA gradient blends down to the point where it goes out of Rec2020 gamut, and then seems to reset back to a darker color and blend down to a lighter color again. When using clipping/no GMA, the Chromium GMA gradient looks fine. Because this affects both MINDE and raytracing, I'm not sure offhand if this is indicative of something with the GMAs themselves, something in the Firefox color translation chain, or maybe in the Chromium GMA implementation in the playground. Any insights you have would be appreciated.

Sample links to test:
* https://apps.colorjs.io/gamut-mapping/gradients?from=oklch%2890%25+.1+20%29&to=oklch%2890%25+.2+20%29
* https://apps.colorjs.io/gamut-mapping/gradients?from=oklch%2880%25+0.1+20%29&to=oklch%2890%25+.2+20%29

It's most noticeable when taking the max DeltaE between steps down to 1.

What I see here (respective order of the bullets above):

<img width="1356" height="135" alt="Image" src="https://github.com/user-attachments/assets/75c777c1-4e6a-4934-97fc-936e90b35162" />

<img width="1364" height="133" alt="Image" src="https://github.com/user-attachments/assets/89c4826e-ab9b-4aeb-9616-a79c62d90612" />

The break in the gradient happens where the color goes out of Rec2020 gamut. Let me know if any more info might be helpful, or if this should go in a separate discussion.

**Comment by jamesnw** [comment]
> One thing I noticed in the gradient GM playground here that I'm wondering if either of you see on your sides when using Nightly: when enabling either local MINDE or the fixed raytracing GMA, the Chromium GMA gradient blends down to the point where it goes out of Rec2020 gamut, and then seems to reset back to a darker color and blend down to a lighter color again. When using clipping/no GMA, the Chromium GMA gradient looks fine. Because this affects both MINDE and raytracing, I'm not sure offhand if this is indicative of something with the GMAs themselves, something in the Firefox color translation chain, or maybe in the Chromium GMA implementation in the playground. Any insights you have would be appreciated.

I did notice this as well. The Chromium GMA listed here was a POC that was first implemented in Chromium, but not released. I reimplemented it in ColorJS as a comparison. IIRC, it gets close to the Rec2020 gamut, but makes some approximations for speed that mean some things are out of the Rec2020 gamut. I think what is happening is that those specific colors are hitting the f32 issue- applying two GMAs is going to produce odd results if the initial one doesn't map inside the gamut. I don't think this is a cause of concern.

**Comment by facelessuser** [comment]
> it looks like the `cast_ray` comparison epsilon/threshold value for f32s is the culprit.

I'm glad the solution was straightforward. I initially looked at the implementation and was confused, as everything looked spot on in terms of implementation. Luckily, I did finally catch on to the different unit types.

I honestly didn't know what unit types browsers were going to use, so it hadn't occurred to me to optimize for a specific unit type. It's likely that even if 1e-6 was used in float64, the results would still be good enough.

If you run into issues like this in the future, you could always use 1e-5 to get a little more headroom. Losing precision on really small numbers isn't really an issue; it's only when you have to invert that same value (1 / x) and multiply by it that it becomes a problem later, as then you are using a very large number with loss of precision. That's where the breakdown happens. It's likely that in all practical senses, 1e-6 will be fine, but something to note is that if real-world use pushes things a little too hard into observable issues, relaxing to 1e-5 might be okay.

**Comment by jparnzen** [comment]
@jamesnw : thanks for the clarification on the Chromium GMA implementation in color.js! And glad to know it's not a showstopper here--that's what I was hoping for :) Let me or the Firefox team know if you run into any other cases where either Firefox GMA shows weirdness.

@facelessuser : I'm glad it was a simple one-and-done bug too. Thanks for catching that again. I've been implementing the GMAs separately in my own projects and hadn't considered the precision difference since mine use f64 just like ColorAide, color.js, and the draft assumption. Good to know for future work in Firefox since the CSS code uses f32. And I'll add a note to the raytracing code when I submit some reftests to Firefox hopefully later today.

**Comment by romainmenke** [comment]
@jparnzen Super cool to see an implementation! Thank you for working on this 🙇 
Is it correct that gamut mapping in the Firefox Nightly always uses `srgb` as the destination and not `rec2020` (https://github.com/w3c/csswg-drafts/issues/9449) or the device gamut?

**Comment by jparnzen** [comment]
@romainmenke : You are correct. To my understanding, Firefox uses sRGB as the destination display color space, and that's the destination space I pass to the GMAs. I saw in some earlier discussions here that there was mention of using Rec2020 during gamut mapping, but I don't think I saw it mentioned in the GMAs themselves in the drafts. Please correct me if I'm wrong.

And many thanks! It's been fun getting these implemented, and great to be able use them and OKLCH in the browser without having to go through an extra layer of a library or manual calculation. And very happy to help the team here work through any  implementation details to make the spec more bullet-proof.

BTW, I submitted the fix for the raytracing bugs that were found, and it autolanded in Nightly on 06-09. Not sure if it's gone into the main branch of Nightly yet or not, but should very soon. I'm working on WPT tests for it now. And if it helps you all out, I added initial WPT tests for local MINDE gamut mapping as part of the GMA implementations; I started raytracing WPTs but the patch landed before I could finish them at the time :)

**Comment by facelessuser** [comment]
> I saw in some earlier discussions here that there was mention of using Rec2020 during gamut mapping, but I don't think I saw it mentioned in the GMAs themselves in the drafts.

That was for a proposal of gamut mapping to Rec2020 and then clipping to the display gamut instead of gamut mapping directlytly to the display gamut. That is not specifically part of the Raytrace or MINDE algorithm (though they could be used to do that). I believe that sRGB, Display P3, and Rec. 2020 are the only supported display gamuts in CSS.

I'm not sure if Firefox yet displays colors in the display gamut (by default at least). I know on my Display P3 monitor that only Safari and Chrome show colors in my display's native capability, Firefox has always only clipped everything to sRGB, but maybe there is a setting that would allow it to 🤷🏻.

**Comment by romainmenke** [comment]
There is a compromise to be made between wanting to show as many colors as possible on any given device and avoiding clipping. And at the same time another compromise between showing as many colors on any device and having consistency between various devices.

## Gamut mapping to `srgb`:

_as currently implemented in Firefox_

- avoid clipping entirely on devices that can display all of `srgb`
- makes it impossible to display wider gamut colors (display-p3 and beyond) when using notations like `oklch`

This forces authors to pick between wider gamut colors and gradients/color transforms in perceptually uniform color spaces.

If we assume that most devices can display all or most of `srgb` then this method does give consistency between devices.

I don't think this is a good compromise. Especially as screens continue to improve and move beyond `srgb`.

## Gamut mapping to `display-p3`, `rec2020`, ...:

_there is a resolution to always gamut map to `rec2020`_

This has the same issues as gamut mapping to `srgb`. But will result in more clipping while making it possible to also display wide gamut colors.

Exactly how much clipping or mapping you get depends on the device.

I think this is marginally better than gamut mapping to `srgb` but that this not what authors want. The unpredictability of when you get clipping and what the outcome of that clipping will be just isn't great.

## Gamut mapping to what the device can actually display

- removes all clipping entirely
- allows wider gamut colors and will display more of them over time as displays become better

Given that clipped colors tend have a larger difference from the original than gamut mapped colors it should also make colors more consistent between devices if we always gamut map to what the device is capable of.

(this last one is my preferred method)

-----

@jparnzen Is it possible to add another flag in Firefox to control the destination color space?
Can this be a toggle between:
- `srgb`
- `display-p3`
- `rec2020`
- the actual display/device

This will give better insights on what is the best approach 🙇

**Comment by jparnzen** [comment]
@romainmenke : as of today, from what I see in the code, Firefox hardcodes queries to get the screen color gamut to return sRGB (in the nsScreen and Screen widget code, for reference).

I think I may have seen a ticket in bugzilla to add support for P3 and maybe Rec2020, but I'll have to go back and look, and I don't think it's gotten any traction as yet. I will say that as part of my work on the GMAs, the function/method signatures for the GMAs take a destination color gamut, with the anticipation that someone will want to pass something other than sRGB at some point.

Thanks for the analysis and breakdown of the 3 possible approaches! I will offer that as a designer and dev, I expect that the colors I specify in CSS should display as close as possible to my expectation, especially with lightness. So clipping to sRGB from a wider gamut space wouldn't sit well, unless the difference between clipped and mapped-to-screen-gamut values are within some unnoticeable threshold. Colors with high lightness levels were a big reason I started on the GMA implementations--trying to get colors at 95%-99% lightness (with some sense of chroma or saturation) in OKLCH to use as things like backgrounds or text color on dark backgrounds is problematic without the GMAs. And while I agree with your take and would like the colors mapped to end-display gamuts, at the very least getting them to sRGB for now seems acceptable as that's effectively the baseline for monitors (I still found monitors for sale with less than 98% sRGB coverage when shopping for a new one recently). Just my anecdotal take :)

**Comment by LeaVerou** [comment]
Btw I just added a [benchmark](https://apps.colorjs.io/gamut-mapping/benchmark/) that compares the existing GMAs for both accuracy and speed, using min/max/avg/median error (which is a weighed sum of l, c, h deltas weighing ΔH 8x, ΔL 4x, and ΔC 1x). It also allows customizing the range and granularity of the patches via URL params.

**I also added a new GMA called OKLCh Cubic**. It is based on the observation that at fixed L and H each linear-P3 channel is exactly a cubic equation in chroma (l' is affine in C, then cubed), so the C boundary is the smallest chroma at which any channel reaches 0 or 1 — found via standard closed form cubic equation solving formulas and optimized further through some observations about the types of equations produced and a lazily populated hue cache (which is not necessary, just an additional optimization).

Because of this:
- it is much simpler than the other leading methods ([< 150 loc](https://github.com/color-js/apps/blob/main/gamut-mapping/methods/oklch-cubic.js))
- does not depend on a LUT
- it is not iterative

In the benchmark it consistently scores **at the top for all accuracy metrics** (min, max, avg, med, iqm) alongside EdgeSeeker, and seems to be 5-30% faster than EdgeSeeker, both of which are even faster than clip!

<img width="1793" height="847" alt="Image" src="https://github.com/user-attachments/assets/37951716-9ee2-48d3-927c-0359813751a3" />

**Comment by facelessuser** [comment]
This is very interesting. But does this only hold true for P3 and not sRGB or Rec. 2020? That would, unfortunately, limit it to P3 only.

Small note, it seems the algorithm kicks out if `!(C > 0)` and doesn't check lightness at all. This would need to be fixed.

I do think these benchmarks don't completely tell the whole story. The way it is being tested is highly advantageous to the cubic approach as it bypasses all conversions; it never leaves the OkLCh color space. That doesn't discredit its speed, only that the extreme advantage it is showing comes specifically from never having to convert out of OkLCh. Give it a P3 color to start, and then it must incur the conversion cost, dramatically dropping its performance gain. It's still very fast, but not 30% faster. But, if you are in OkLCh, technically, it would be even faster as it skips the conversions. I think in browsers, we would always be converting to the actual gamut at the end, though, for display.

<img width="262" height="660" alt="Image" src="https://github.com/user-attachments/assets/c7ec02a6-118a-4bbf-b75e-6001812ba22c" />

**Comment by facelessuser** [comment]
I've updated the algorithms for a fairer comparison. GMAs all return the gamut mapped colors consistently in the origin space. I also fixed the oklch-cubic method so that it actually checks lightness for gamut mapping. https://github.com/color-js/apps/pull/36#issuecomment-4754910626.

OkLCh-cubic is still usually the fastest, but results seem more fair now, instead of each algorithm returning the color in inconsistent color spaces.

<img width="247" height="654" alt="Image" src="https://github.com/user-attachments/assets/2da8944f-7cf2-4bbe-8682-0300a6b9b0c1" />

**Comment by facelessuser** [comment]
I implemented the new cubic algorithm on an experimental branch in the ColorAide library to see how it performed in a real-world environment with similar overhead and abstraction for generalization within the library. 

Compared against Ray Trace (as I don't have the LUT method implemented), it was the fastest by far.

This was processing an image with varying ranges of lightness and hue with maximum saturation in Rec. 2020. We gamut-mapped them to sRGB.

```py
➜  coloraide git:(feature/fit-oklch-cubic) ✗ python3 tools/apply_gamut.py -i test.jpg -o test-map4.jpg --gamut rec2020 --gmap 'raytrace'
Pixels: 1048576
> 100%
Completed in: 7.003060333 sec
➜  coloraide git:(feature/fit-oklch-cubic) ✗ python3 tools/apply_gamut.py -i test.jpg -o test-map4.jpg --gamut rec2020 --gmap 'oklch-cubic'
Pixels: 1048576
> 100%
Completed in: 3.657383834 sec
```

I didn't see it overpower clip in my library, and that may be for any number of design reasons, but Ray Trace vs Cubic was the most apples-to-apples comparison when it came to generalized abstraction and overhead. Admittedly, our version of Ray Trace has some extra stuff outside of what's specified in the CSS spec that adds a little more overhead, but even if we were to strip those out, there is no way it could overtake the Cubic approach.

Generally, it also provided fewer deviations in lightness and hue. While the difference wasn't visibly noticeable, the fact that it can reduce the chroma that fast and more accurately is impressive.

**Comment by facelessuser** [comment]
After abstracting the OkLCh Cubic algorithm to work for any linear RGB color space, and evaluating it, I do think it is worth considering, as it seems fast and accurate for sRGB, Display P3, and Rec. 2020.

With or without the additional hue data cache, it is still faster than Ray Trace. If a cache were used, I would probably think having some bounds would make sense, but I think the algorithm is extremely competitive even without the cache. So if keeping a LUT of hue data was not desired, it is still very fast and accurate.

I do not think it is actually faster than clip, and further, I would state I believe it needs a clip itself at the end of the algorithm, as it can produce some values that do need clipping in some circumstances.

I found it impressive enough to implement it and release it in ColorAide as an experimental GMA to play with.

**Comment by facelessuser** [comment]
As a side note, I've issued a PR with some corrections to the Ray Trace algorithm: https://github.com/w3c/csswg-drafts/pull/14080. It addresses:

1. It addresses the issue [here](https://github.com/w3c/csswg-drafts/issues/10579#issuecomment-4668607031) and now notes values for both 64-bit and 32-bit.
2. It addresses the issues noted [here](https://github.com/w3c/csswg-drafts/issues/14029).
3. And one other improvement is that it notes optimal values for the distance threshold from the gamut surface. Previously, this was specified as 1e-6, and it has been updated to 1e-12 (for 64-bit) and noted that 1e-6 should be used for 32-bit. This reduces hue deltas in some cases.

**Comment by facelessuser** [comment]
Small update to the benchmarks that were posted earlier using ColorAide. The ColorAide version allows swapping out OkLCh for any perceptual color space (as the approach can be used for most other perceptual spaces).

Here's a fairer comparison with a Ray Trace purpose-built for OkLCh and linear CSS RGB spaces. The Cubic one is still shown to be faster, but the performance is much closer. Now

```console
➜  coloraide git:(main) ✗ python3 tools/benchmark_gma.py -t oklch -m 'raytrace-oklch'
Colors: 250000
> 100%
Completed in: 4.91770125 sec
➜  coloraide git:(main) ✗ python3 tools/benchmark_gma.py -t oklch -m 'oklch-cubic'
Colors: 250000
> 100%
Completed in: 4.384709042 sec
```

Now, add the cache, and the Cubic approach pulls further ahead:

```console
➜  coloraide git:(main) ✗ python3 tools/benchmark_gma.py -t oklch -m 'oklch-cubic:{"cache": true}'
Colors: 250000
> 100%
Completed in: 3.021533917 sec
```

**Comment by ardov** [comment]
I added another LUT-less algorithm based on bracketed Halley method (https://github.com/color-js/apps/pull/44#issuecomment-4952409409)

It is similar to OKLCh cubic but there are a few differences:
- It uses Halley iterations which seem to work faster (not using expensive acos/cbrt). In Color JS benchmark it's usually performs 10-20% faster than OKLCh cubic.
- It takes into account the problematic "blue fold" that happens in sRGB and Rec.2020. Algorithms like binary search tend to jump over more vivid blues and return noticeably duller colors. Originally, it was the main reason for me to create the Edge Seeker.

<img width="1192" height="547" alt="Image" src="https://github.com/user-attachments/assets/0e336d96-b554-4f44-942d-785a595f73d2" />

---

Oh and I also wanted to share the quality measuring approach I used for the algorithms.

The problem with a random out-of-gamut color is that we don't have a ground truth to compare with. 
So I took RGB colors on the surface of the gamut → converted to OKLCh → increased chroma. I was feeding the algorithm with the increased-chroma versions. And then I compared the results with the known original RGB colors to see how well it performed. I did it for all ~320K hexes on the surface of the gamut.

**Comment by svgeesus** [comment]
Review welcome on a PR to add Edge Seeker pseudocode (it *seems* correct but I would value other eyes on it):

 - https://github.com/w3c/csswg-drafts/pull/14426
